memtagabielf64/memtagabielf64.rst

```diff
@@ -0,0 +1,723 @@
+designed to be backwards-compatible, the result of the relocation on a
+non-MemtagABI binary or non-MTE device is the same as the non-extended variant.
+
+The relocations reference the following mnemonics:
+
+- ``LDG(value)`` is an instruction for the run-time environment to materialize a
+  tagged pointer from the provided address, using the ``ldg`` instruction. This
+  operation should also align the pointer down to the closest tag granule before
+  executing the ``ldg`` instruction.
+
+- ``TAG(pointer, offset)`` is an instruction for the run-time environment to
+  fetch the memory tag at ``(pointer + offset) & ~0xf``, which can be done using
+  the ``ldg`` instruction. Then, this tag should be materialized to ``pointer``,
+  overwriting the existing address tag in ``pointer`` if it exists.
+
+- For all the relocation types listed below, the loader should use the ``ldg``
+  instruction on the target address before writing into the target field, as the
+  target field may be inside of a tagged region.
+
+.. table:: Relocations with extended semantics
+
+  +------------+--------------------+----------------+-------------------------------------+
+  | ELF64 Code | Name               | Base Operation | MemtagABI Extended Operation        |
+  +============+====================+================+=====================================+
+  | 257	      | R_AARCH64_ABS64    | S + A          | LDG(S) + A                          |
+  +------------+--------------------+----------------+-------------------------------------+
+  | 1025	      | R_AARCH64_GLOB_DAT | S + A          | LDG(S) + A                          |
+  +------------+--------------------+----------------+-------------------------------------+
+  | 1027	      | R_AARCH64_RELATIVE | Delta(S) + A   | TAG(Delta(S) + A, \*(Delta(S) + A)) |
+  +------------+--------------------+----------------+-------------------------------------+
+
+``R_AARCH64_ABS64`` and ``R_AARCH64_GLOB_DAT`` are thus extended to materialize
+the memory tag of the symbol ``S`` for relocation purposes.
+
+Extended semantics of ``R_AARCH64_RELATIVE``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``R_AARCH64_RELATIVE`` is extended with special care to preserve backwards
+compatibility. This relocation extension allows for the out-of-bounds derived
+symbols to be resolved with the correct memory tags, as shown in the example
+below:
+
+.. code-block::
+
+   // Comments show the resulting dynamic relocations if this file were
+   // compiled with -fPIC.
+
+   // GOT_ENTRY_FOR_foo populated with R_AARCH64_RELATIVE, as usual, with
+   // addend zero and tag-offset zero. This can be compressed to a RELR
+   // relocation.
+   int* foo[32];
+
+   // GOT_ENTRY_FOR_foo_middle populated with R_AARCH64_RELATIVE, with addend
+   // 16 and tag-offset -16. Because the linker knows that the tag-offset is
+   // in-bounds of the symbol, the linker can omit the tag-offset, and
+   // compress this to a RELR relocation.
```

Comment:
Would be good to add a reference to RELR in the references section. Best that I know of is https://groups.google.com/g/generic-abi/c/bX460iggiKg